File: src/main/java/com/robohorse/robopojogenerator/generator/processing/ClassProcessor.java
package com.robohorse.robopojogenerator.generator.processing;

import com.robohorse.robopojogenerator.generator.consts.ClassEnum;
import com.robohorse.robopojogenerator.generator.consts.common.ClassField;
import com.robohorse.robopojogenerator.generator.consts.common.ClassItem;
import com.robohorse.robopojogenerator.generator.consts.common.JsonModel;
import com.robohorse.robopojogenerator.generator.consts.common.JsonModel.JsonItem;
import com.robohorse.robopojogenerator.generator.consts.common.JsonModel.JsonItemArray;
import com.robohorse.robopojogenerator.generator.consts.templates.ImportsTemplate;
import com.robohorse.robopojogenerator.generator.utils.ClassGenerateHelper;
import org.json.JSONArray;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by vadim on 23.09.16.
 */
public class ClassProcessor {
    ClassGenerateHelper classGenerateHelper;

    public ClassProcessor(ClassGenerateHelper classGenerateHelper) {
        this.classGenerateHelper = classGenerateHelper;
    }

    public void proceed(JsonModel jsonItem, final Map<String, ClassItem> itemMap) {
        final ClassItem classItem = new ClassItem(classGenerateHelper.formatClassName(jsonItem.getKey()));
        if (jsonItem instanceof JsonItem) {
            processSingleObject((JsonItem) jsonItem, itemMap, classItem);
        } else {
            processArrayObject((JsonItemArray) jsonItem, itemMap, classItem);
        }
        appendItemsMap(itemMap, classItem);
    }

    private void processArrayObject(JsonItemArray arrayItem,
                                    final Map<String, ClassItem> itemMap,
                                    final ClassItem classItem) {
        final JSONArray jsonArray = arrayItem.getJsonObject();
        classItem.getClassImports().add(ImportsTemplate.LIST);

        final ClassField classField = new ClassField();
        if (jsonArray.length() == 0) {
            classField.setClassField(new ClassField(ClassEnum.OBJECT));
            classItem.getClassFields().put(arrayItem.getKey(), classField);
        } else {
            final JsonItemArray jsonItemArray = new JsonItemArray(arrayItem.getKey(), jsonArray);
            proceedArray(jsonItemArray, classField, itemMap);
            classItem.getClassFields().put(arrayItem.getKey(), classField);
        }
    }

    private void processSingleObject(JsonItem jsonItem,
                                     final Map<String, ClassItem> itemMap,
                                     final ClassItem classItem) {
        for (final String jsonObjectKey : jsonItem.getJsonObject().keySet()) {
            final Object object = jsonItem.getJsonObject().get(jsonObjectKey);
            final InnerObjectResolver innerObjectResolver = new InnerObjectResolver() {

                @Override
                public void onInnerObjectIdentified(ClassEnum classType) {
                    classItem.getClassFields().put(jsonObjectKey, new ClassField(classType));
                }

                @Override
                public void onJsonObjectIdentified() {
                    final String className = classGenerateHelper.formatClassName(jsonObjectKey);
                    final ClassField classField = new ClassField(null, className);
                    final JsonItem jsonItem = new JsonItem(jsonObjectKey, (JSONObject) object);

                    classItem.getClassFields().put(jsonObjectKey, classField);
                    proceed(jsonItem, itemMap);
                }

                @Override
                public void onJsonArrayIdentified() {
                    final JSONArray jsonArray = (JSONArray) object;
                    classItem.getClassImports().add(ImportsTemplate.LIST);

                    final ClassField classField = new ClassField();
                    if (jsonArray.length() == 0) {
                        classField.setClassField(new ClassField(ClassEnum.OBJECT));
                        classItem.getClassFields().put(jsonObjectKey, classField);

                    } else {
                        final JsonItemArray jsonItemArray = new JsonItemArray(jsonObjectKey, (JSONArray) object);
                        proceedArray(jsonItemArray, classField, itemMap);
                        classItem.getClassFields().put(jsonObjectKey, classField);
                    }
                }
            };
            innerObjectResolver.resolveClassType(object);
        }
    }

    private void appendItemsMap(Map<String, ClassItem> itemMap, ClassItem classItem) {
        final String className = classItem.getClassName();
        if (itemMap.containsKey(className)) {
            final ClassItem storedClassItem = itemMap.get(className);
            classItem.getClassFields().putAll(storedClassItem.getClassFields());
        }
        itemMap.put(className, classItem);
    }

    private void proceedArray(final JsonItemArray jsonItemArray,
                              final ClassField classField,
                              final Map<String, ClassItem> itemMap) {
        final String itemName = classGenerateHelper.getClassNameWithItemPostfix(jsonItemArray.getKey());
        if (jsonItemArray.getJsonObject().length() != 0) {
            final Object object = jsonItemArray.getJsonObject().get(0);
            final InnerObjectResolver innerObjectResolver = new InnerObjectResolver() {

                @Override
                public void onInnerObjectIdentified(ClassEnum classType) {
                    classField.setClassField(new ClassField(classType));
                }

                @Override
                public void onJsonObjectIdentified() {
                    final int size = jsonItemArray.getJsonObject().length();
                    final Map<String, ClassItem> innerItemsMap = new HashMap<String, ClassItem>();
                    for (int index = 0; index < size; index++) {
                        final JSONObject jsonObject = (JSONObject) jsonItemArray.getJsonObject().get(index);
                        final JsonItem jsonItem = new JsonItem(itemName, jsonObject);
                        proceed(jsonItem, innerItemsMap);
                    }
                    classField.setClassField(new ClassField(null, itemName));
                    innerItemsMap.forEach((key, value) -> {
                        ClassItem existing = itemMap.get(key);
                        if (existing != null) {
                            value.getClassFields().forEach((classKey, classValue) -> existing.getClassFields().putIfAbsent(classKey, classValue));
                            existing.getClassImports().addAll(value.getClassImports());
                        } else {
                            itemMap.put(key, value);
                        }
                    });
                }

                @Override
                public void onJsonArrayIdentified() {
                    classField.setClassField(new ClassField());
                    final JsonItemArray jsonItemArray = new JsonItemArray(itemName, (JSONArray) object);
                    proceedArray(jsonItemArray, classField, itemMap);
                }
            };
            innerObjectResolver.resolveClassType(object);

        } else {
            classField.setClassField(new ClassField(ClassEnum.OBJECT));
        }
    }
}

File: generator/src/main/kotlin/com/robohorse/robopojogenerator/generator/processing/ClassProcessor.kt
package com.robohorse.robopojogenerator.generator.processing

import com.robohorse.robopojogenerator.generator.common.ClassEnum
import com.robohorse.robopojogenerator.generator.common.common.ClassField
import com.robohorse.robopojogenerator.generator.common.common.ClassItem
import com.robohorse.robopojogenerator.generator.common.common.JsonModel
import com.robohorse.robopojogenerator.generator.common.common.JsonModel.JsonItemArray
import com.robohorse.robopojogenerator.generator.common.templates.ImportsTemplate
import com.robohorse.robopojogenerator.generator.utils.ClassGenerateHelper
import org.json.JSONArray
import org.json.JSONObject

internal class ClassProcessor(
    private val classGenerateHelper: ClassGenerateHelper
) {

    fun proceed(jsonItem: JsonModel, itemMap: LinkedHashMap<String?, ClassItem>) {
        val classItem = ClassItem(classGenerateHelper.formatClassName(jsonItem.key))
        if (jsonItem is JsonModel.JsonItem) {
            processSingleObject(jsonItem, itemMap, classItem)
        } else {
            processArrayObject(jsonItem as JsonItemArray, itemMap, classItem)
        }
        appendItemsMap(itemMap, classItem)
    }

    private fun processArrayObject(
        arrayItem: JsonItemArray,
        itemMap: MutableMap<String?, ClassItem>,
        classItem: ClassItem
    ) {
        val jsonArray = arrayItem.jsonObject
        classItem.classImports.add(ImportsTemplate.LIST)
        val classField = ClassField()
        if (jsonArray.length() == 0) {
            classField.classField = ClassField(ClassEnum.OBJECT)
            classItem.classFields.put(arrayItem.key, classField)
        } else {
            val jsonItemArray = JsonItemArray(arrayItem.key, jsonArray)
            proceedArray(jsonItemArray, classField, itemMap)
            classItem.classFields.put(arrayItem.key, classField)
        }
    }

    private fun processSingleObject(
        jsonItem: JsonModel.JsonItem,
        itemMap: LinkedHashMap<String?, ClassItem>,
        classItem: ClassItem
    ) {
        for (jsonObjectKey in jsonItem.jsonObject.keySet()) {
            val `object` = jsonItem.jsonObject[jsonObjectKey]
            val innerObjectResolver: InnerObjectResolver = object : InnerObjectResolver() {
                override fun onInnerObjectIdentified(classType: ClassEnum?) {
                    classItem.classFields.put(jsonObjectKey, ClassField(classType))
                }

                override fun onJsonObjectIdentified() {
                    val className = classGenerateHelper!!.formatClassName(jsonObjectKey)
                    val classField = ClassField(null, className)
                    val jsonItem = JsonModel.JsonItem(jsonObjectKey, (`object` as JSONObject))
                    classItem.classFields.put(jsonObjectKey, classField)
                    proceed(jsonItem, itemMap)
                }

                override fun onJsonArrayIdentified() {
                    val jsonArray = `object` as JSONArray
                    classItem.classImports.add(ImportsTemplate.LIST)
                    val classField = ClassField()
                    if (jsonArray.length() == 0) {
                        classField.classField = ClassField(ClassEnum.OBJECT)
                        classItem.classFields.put(jsonObjectKey, classField)
                    } else {
                        val jsonItemArray = JsonItemArray(
                            jsonObjectKey,
                            `object`
                        )
                        proceedArray(jsonItemArray, classField, itemMap)
                        classItem.classFields.put(jsonObjectKey, classField)
                    }
                }
            }
            innerObjectResolver.resolveClassType(`object`)
        }
    }

    private fun appendItemsMap(itemMap: MutableMap<String?, ClassItem>, classItem: ClassItem) {
        val className = classItem.className
        if (itemMap.containsKey(className)) {
            val storedClassItem = itemMap[className]
            classItem.classFields.putAll(storedClassItem!!.classFields)
        }
        itemMap[className] = classItem
    }

    private fun proceedArray(
        jsonItemArray: JsonItemArray,
        classField: ClassField,
        itemMap: MutableMap<String?, ClassItem>
    ) {
        val itemName = classGenerateHelper!!.getClassNameWithItemPostfix(jsonItemArray.key)
        if (jsonItemArray.jsonObject.length() != 0) {
            val `object` = jsonItemArray.jsonObject[0]
            val innerObjectResolver: InnerObjectResolver = object : InnerObjectResolver() {
                override fun onInnerObjectIdentified(classType: ClassEnum?) {
                    classField.classField = ClassField(classType)
                }

                override fun onJsonObjectIdentified() {
                    val size = jsonItemArray.jsonObject.length()
                    val innerItemsMap: LinkedHashMap<String?, ClassItem> = LinkedHashMap()
                    for (index in 0 until size) {
                        val jsonObject = jsonItemArray.jsonObject[index] as JSONObject
                        val jsonItem = JsonModel.JsonItem(itemName, jsonObject)
                        proceed(jsonItem, innerItemsMap)
                    }
                    classField.classField = ClassField(null, itemName)
                    innerItemsMap.forEach { (key: String?, value: ClassItem) ->
                        val existing = itemMap[key]
                        if (existing != null) {
                            value.classFields.forEach { (classKey: String?, classValue: ClassField?) ->
                                existing.classFields.putIfAbsent(
                                    classKey,
                                    classValue
                                )
                            }
                            existing.classImports.addAll(value.classImports)
                        } else {
                            itemMap[key] = value
                        }
                    }
                }

                override fun onJsonArrayIdentified() {
                    classField.classField = ClassField()
                    val jsonItemArray = JsonItemArray(itemName, (`object` as JSONArray))
                    proceedArray(jsonItemArray, classField, itemMap)
                }
            }
            innerObjectResolver.resolveClassType(`object`)
        } else {
            classField.classField = ClassField(ClassEnum.OBJECT)
        }
    }
}